/*! This file is part of davids91/Rafko.
 *
 *    Rafko is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    Rafko is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with Rafko.  If not, see <https://www.gnu.org/licenses/> or
 *    <https://github.com/davids91/rafko/blob/master/LICENSE>
 */

syntax = "proto3";
option cc_enable_arenas = true;
option java_package = "org.rafko.mainframe";
option java_outer_classname = "RafkoDeepLearningService";

package rafko_mainframe;

import "common.proto";
import "sparse_net.proto";

/** @brief      A service consists of slots on which different Neural related tasks can run.
 *              The differetn kinds of tasks are contained in the enum @Slot_type
 */
enum Slot_type{
  SERV_SLOT_UNKNOWN = 0;
  SERV_SLOT_TO_RUN = 1; /* Simply runs a network, provides output on demand. Ignores @data_set and @test_set in @Service_slot */
  SERV_SLOT_TO_OPTIMIZE = 2; /* Optimizes the attached network in it's loop based on the custom implementation of the Stochastic Gradient descent algorithm */
  SERV_SLOT_TO_APPROXIMIZE = 3; /* Approximizes the attached network in it's loop based on an approximation of its gradient from the provided data sets */
}

/** @brief      A service slot can have different states, all of which are contained in the below enum.
 *              The states shall be summarized in a unit32 field as a bitstring, except when everything is okay,
 *              in which case the value shall be SERV_SLOT_OK
 */
enum Slot_state_values{
  SERV_SLOT_STATE_UNKNOWN = 0;
  SERV_SLOT_MISSING_NET = 1; /* Missing or empty Neural network attached to the Server slot */
  SERV_SLOT_MISSING_SOLUTION = 2; /* Despite Network state, a Solution was not built based on the attached network ( if any ) */
  SERV_SLOT_MISSING_DATA_SET = 4;
  SERV_SLOT_MISSING_COST_FUNCTION = 16;
  SERV_SLOT_MISSING_TRAINER = 32;
  SERV_SLOT_OK = 65536;
}

/** @brief      Available actions to request from a service slot
 */
enum Slot_action_field{
  SERV_SLOT_NOOP = 0;
  SERV_SLOT_TO_START = 1; /* This action makes the server call the @loop of the @target_slot_id every time its @loop is called */
  SERV_SLOT_TO_STOP = 2; /* This action disables the behavior introduced by @SERV_SLOT_TO_START */
  SERV_SLOT_TO_RESET = 4; /* Depending on the slot, it could reset the attached networks data, the test/training error scores or whatever is implemented in the slot's @reset function */
  SERV_SLOT_TO_TAKEOVER_NET = 8; /* Takes a copy from the source slot, and overwrites the network in the target slot in it */
  SERV_SLOT_TO_APPEND_TRAINING_SET = 16; /* Adds the provided data_stream as an element of the training set attached to the server slot */
  SERV_SLOT_TO_APPEND_TEST_SET = 32; /* Adds the provided data_stream as an element of the test set attached to the server slot */
  SERV_SLOT_TO_DISTILL_NETWORK = 64; /* For the network attached to the slot: removes the zero, or insignificantly small weights with connections. */
  /*!Note: The removal/duplication threshold is based on the @beta parameter in the service hyperparameters */
  SERV_SLOT_TO_AMPLIFY_NETWORK = 128; /* For the network attached to the slot: duplicates the significant neurons. */
  /*!Note: A neuron proves to be significant if the output of it is paired with greater weights, than other neurons.*/
  SERV_SLOT_RUN_ONCE = 256; /* Run network one time. Uses the attached @data_stream of the request */
  SERV_SLOT_TO_GET_TRAINING_SAMPLE = 512; /* Provides a training sample[@request_index] from the training set attached to the target slot: first, the input, then the label packed together */
  SERV_SLOT_TO_GET_TEST_SAMPLE = 1024; /* Provides a training sample[@request_index] from the training set attached to the target slot: first, the input, then the label packed together */
  SERV_SLOT_TO_REFRESH_SOLUTION = 2048; /* Re-generates the solution attached to the server slot based on the attached network */
  SERV_SLOT_TO_DIE = 65536;
}

/** @brief      Available information fields to query from a service slot. The data is given back in the order of the enumerations
 *              e.g.: in case the info (SLOT_INFO_TRAINING_ERROR & SLOT_INFO_TEST_SET_SEQUENCE_NUMBER) is be queried, 
 *                    SLOT_INFO_TRAINING_ERROR is be under index 1
 *                    SLOT_INFO_TEST_SET_SEQUENCE_NUMBER shall be under index 2
 *              e.g.: in case the info (SLOT_INFO_TEST_SET_SAMPLE_NUMBER & SLOT_INFO_TRAINING_SET_SAMPLE_NUMBER) is queried
 *                    SLOT_INFO_TRAINING_SET_SAMPLE_NUMBER is be under index 1
 *                    SLOT_INFO_TEST_SET_SAMPLE_NUMBER shall be under index 2
 */
enum Slot_info_field{
  SLOT_INFO_UNKNOWN = 0;
  SLOT_INFO_ITERATION = 1;
  SLOT_INFO_TRAINING_ERROR = 32;
  SLOT_INFO_TRAINING_SET_SEQUENCE_COUNT = 64;
  /*!Note: values related to training set can go up to 512 */
  SLOT_INFO_TEST_ERROR = 1024;
  SLOT_INFO_TEST_SET_SEQUENCE_COUNT = 2048;
  /*!Note: values related to test set can go up to 16384 */
}

/** @brief      Certain slots may require raw I/O data form the network to operate.
 *              The message @Neural_io_stream provides either Neural network input data
 *              or output data, depending on the recipient. One sample of the packet shall 
 *              arrive in the order they are defined here. i.e.: first the input, then the labels,
 *              then the features. COmponents might be left empty, in which case the corresponding 
 *              descriptor shall remain empty.
 */
message Neural_io_stream{
  /* +++ Size components +++ */
  uint32 sequence_size = 1; /* Number of feature sequenced arrays in a package; sequence lengths for each types of data are set below */
  uint32 input_size = 2; /* Number of floating point numbers that a Neuron network takes as input */
  uint32 label_size = 3; /* Number of floating point numbers that a test or training set stores */
  uint32 feature_size = 4; /* Number of floating point numbers that a Neuron network produced */
  /* --- Size components --- */
  repeated double package = 10; /* The packed data */
}

/** @brief      The meaasge to ask requests from the server. @action, @data_stream may be left blank.
 */
message Slot_request{
  string source_slot_id = 1; /* In case of action requests involving 2 service slots, this field shall be used as the source slot */
  string target_slot_id = 2; /* In case of action requests involving only one slot this field shall be used */
  uint32 request_index = 10; /* In case of datasets, the index of the data sample */
  uint32 request_bitstring = 11; /* Either an action or an info request */
  Neural_io_stream data_stream = 20; /* May be a part of a data-set or a run-once request */
}

/** @brief      Provides the parameters for the rpc @build_network in the service @Rafko_deep_learning
 *              This message is a utility for a rpc which is planned to be in the service temporarily.
 */
message Build_network_request{
  string target_slot_id = 2;
  uint32 input_size = 10;
  double expected_input_range = 11;
  repeated sparse_net_library.transfer_functions allowed_transfers_by_layer = 14;
  repeated uint32 layer_sizes = 15;
}

/** @brief      Requests to a server reply with the following response package. Some fields may
 *              be left blank.
 */
message Slot_response{
  string slot_id = 1; /* A string id to query existing service slots by */
  uint32 slot_state = 10; /* a bitfield representing the status of the slot */
  Neural_io_stream data_stream = 30; /* May be a part of a data-set or a run-once request */
}

/**
 * @brief      This class describes a packet of slot information.
 */
message Slot_info{
  repeated Slot_info_field info_field = 10;
  repeated double info_package = 11;
}

/**
 * @brief      Service hyperparameters describe the context in which deep learning services
 *             operate. It contains all the neccessary set of variables to function in the 
 *             context of deep learning. System relevant parameters like the number of working threads
 *             are decided from the server configuration.
 */
message Service_hyperparameters{
  double step_size = 10;
  uint32 minibatch_size = 11;
  uint32 memory_truncation = 12;

  double alpha = 20;
  double beta = 21;
  double gamma = 22;
  double epsilon = 24; /* very small positive value almost greater, than zero */
  double lambda = 31;
};

/** @brief      The structure of a service slot encapsulating a channel in which deep learning related requests
 *              can be run.
 */
message Service_slot{
  string slot_id = 1; /* A unique identifier for each service slot */
  uint32 state = 2; /* A bitfield storing 0 or 1 at the positions given in enum @Slot_state_values, except when everything is fine, then the value @SERV_SLOT_OK is used */
  Slot_type type = 10; /* Must be set on client side */
  Service_hyperparameters hypers = 21;
  sparse_net_library.SparseNet network = 22;
  sparse_net_library.Data_set test_set = 23;
  sparse_net_library.Data_set training_set = 24;
  sparse_net_library.cost_functions cost_function = 25;
  sparse_net_library.weight_updaters weight_updater = 26;
}

/**
 * @brief      Describes a Rafko Deep Learning Service
 */
service Rafko_deep_learning{
  rpc add_slot(Service_slot) returns (Slot_response) {} /* If possible adds a new slot into the server, then retudrns the status and ID of it */
  rpc update_slot(Service_slot) returns (Slot_response) {} /* Modify existing slot; the ID must be provided */

  rpc ping(Slot_request) returns (Slot_response) {} /* Pings a slot, the response repeats the slot id and state if valid */
  rpc build_network(Build_network_request) returns (Slot_response) {} /* Builds a Neural network with the given parameters into the provided slot */
  rpc request_action(stream Slot_request) returns (stream Slot_response) {} /* Make requests for the target service_slot, and return with the status after each request is completed */
  rpc get_info(Slot_request) returns (Slot_info) {}
  rpc get_network(Slot_request) returns (sparse_net_library.SparseNet) {} /* Provides the Loaded network into the slot */
}